internal/runtime/metrics/metrics.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package metrics provides functions for creating Runtime SDK related metrics.
+package metrics
+
+import (
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	ctrlmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+func init() {
+	// Register the metrics at the controller-runtime metrics registry.
+	ctrlmetrics.Registry.MustRegister(RequestsTotal.metric)
+	ctrlmetrics.Registry.MustRegister(RequestLatency.metric)
+}
+
+// Metrics subsystem and all of the keys used by the Runtime SDK.
+const (
+	runtimeSDKSubsystem = "runtime_sdk"
+	requestLatencyKey   = "request_latency_seconds"
+	requestsTotalKey    = "requests_total"
+)
+
+var (
+	// RequestsTotal reports request result by return code, method, host and hook.
+	RequestsTotal = requestsTotalObserver{
+		prometheus.NewCounterVec(prometheus.CounterOpts{
+			Subsystem: runtimeSDKSubsystem,
+			Name:      requestsTotalKey,
+			Help:      "Number of HTTP requests, partitioned by status code, method, and host.",
+		}, []string{"code", "method", "host", "hook"}),
+	}
+	// RequestLatency reports the request latency in seconds per hook and host.
+	RequestLatency = latencyMetricObserver{
+		prometheus.NewHistogramVec(prometheus.HistogramOpts{
+			Subsystem: runtimeSDKSubsystem,
+			Name:      requestLatencyKey,
+			Help:      "Request latency in seconds, broken down by verb and host.",
+			Buckets:   prometheus.ExponentialBuckets(0.001, 2, 10),
+		}, []string{"hook", "host"}),
+	}
+)
+
+type requestsTotalObserver struct {
+	metric *prometheus.CounterVec
+}
+
+type latencyMetricObserver struct {
+	metric *prometheus.HistogramVec
+}
+
+func (m *requestsTotalObserver) increment(code, method, host, hook string) {
+	m.metric.WithLabelValues(code, method, host, hook).Inc()
+}
+
+// Observe observes a http request result and increments the metric for the given
+// error code, method, host and hook.
+func (m *requestsTotalObserver) Observe(req *http.Request, resp *http.Response, hook string, err error) {
+	host := "none"
```

Comment:
Is this something that can happen? If yes, what about the latency metric?

Reply:
This is some prior art from rest/request.go, seems like they slighty changed it in the past.
I think because we define the urls in our code it is fine to bypass that.